File: core/gamification.py
"""Gamification: XP, hearts, streaks, levels."""

from datetime import date, datetime, timedelta

from core.database import get_profile, update_profile, get_today_stats, update_today_stats


# XP rewards
XP_CORRECT = 10
XP_INCORRECT = 5  # effort reward
XP_LESSON_BONUS = 50
XP_STREAK_BONUS = 20


def award_xp(correct: bool) -> int:
    """Award XP for answering an exercise. Returns XP earned."""
    xp = XP_CORRECT if correct else XP_INCORRECT
    profile = get_profile()
    new_total = profile["total_xp"] + xp
    new_level = (new_total // 100) + 1
    update_profile(total_xp=new_total, level=new_level)
    update_today_stats(xp_delta=xp, exercises_delta=1)
    return xp


def complete_lesson() -> int:
    """Award lesson completion bonus. Returns total bonus XP."""
    profile = get_profile()
    bonus = XP_LESSON_BONUS

    # Streak bonus
    today_stats = get_today_stats()
    if not today_stats["streak_maintained"]:
        bonus += XP_STREAK_BONUS
        _update_streak()

    new_total = profile["total_xp"] + bonus
    new_level = (new_total // 100) + 1
    update_profile(total_xp=new_total, level=new_level)
    update_today_stats(xp_delta=bonus, lesson_completed=True)
    return bonus


def _update_streak() -> None:
    """Update streak counter."""
    profile = get_profile()
    yesterday = (date.today() - timedelta(days=1)).isoformat()

    from core.database import get_connection
    conn = get_connection()
    yesterday_row = conn.execute(
        "SELECT streak_maintained FROM daily_stats WHERE date = ?", (yesterday,)
    ).fetchone()
    conn.close()

    if yesterday_row and yesterday_row["streak_maintained"]:
        new_streak = profile["current_streak"] + 1
    else:
        new_streak = 1

    longest = max(profile["longest_streak"], new_streak)
    update_profile(current_streak=new_streak, longest_streak=longest)


def use_heart() -> bool:
    """Consume a heart on wrong answer. Returns False if no hearts left."""
    from core.database import recharge_hearts
    hearts = recharge_hearts()
    if hearts <= 0:
        return False
    update_profile(hearts=hearts - 1)
    return True


def get_hearts() -> int:
    from core.database import recharge_hearts
    return recharge_hearts()


def get_level_info() -> dict:
    profile = get_profile()
    level = profile["level"]
    xp = profile["total_xp"]
    xp_in_level = xp % 100
    return {
        "level": level,
        "total_xp": xp,
        "xp_in_level": xp_in_level,
        "xp_to_next": 100 - xp_in_level,
    }

File: classlingo-v2/src/lib/engine/gamification.ts
/**
 * Gamification: XP, hearts, streaks, levels.
 * Port of Python core/gamification.py
 */

import { SupabaseClient } from "@supabase/supabase-js";
import {
  getProfile,
  updateProfile,
  getTodayStats,
  updateTodayStats,
  rechargeHearts,
} from "@/lib/db/queries";

// XP rewards
export const XP_CORRECT = 10;
export const XP_INCORRECT = 5; // effort reward
export const XP_LESSON_BONUS = 50;
export const XP_STREAK_BONUS = 20;

/** Award XP for answering an exercise. Returns XP earned. */
export async function awardXp(supabase: SupabaseClient, correct: boolean): Promise<number> {
  const xp = correct ? XP_CORRECT : XP_INCORRECT;
  const profile = await getProfile(supabase);
  if (!profile) return 0;

  const newTotal = profile.total_xp + xp;
  const newLevel = Math.floor(newTotal / 100) + 1;
  await updateProfile(supabase, { total_xp: newTotal, level: newLevel });
  await updateTodayStats(supabase, { xp_delta: xp, exercises_delta: 1 });
  return xp;
}

/** Award lesson completion bonus. Returns total bonus XP. */
export async function completeLesson(supabase: SupabaseClient): Promise<number> {
  const profile = await getProfile(supabase);
  if (!profile) return 0;

  let bonus = XP_LESSON_BONUS;

  const todayStats = await getTodayStats(supabase);
  if (todayStats && !todayStats.streak_maintained) {
    bonus += XP_STREAK_BONUS;
    await updateStreak(supabase);
  }

  const newTotal = profile.total_xp + bonus;
  const newLevel = Math.floor(newTotal / 100) + 1;
  await updateProfile(supabase, { total_xp: newTotal, level: newLevel });
  await updateTodayStats(supabase, { xp_delta: bonus, lesson_completed: true });
  return bonus;
}

async function updateStreak(supabase: SupabaseClient): Promise<void> {
  const profile = await getProfile(supabase);
  if (!profile) return;

  const { data: { user } } = await supabase.auth.getUser();
  if (!user) return;

  const yesterday = new Date();
  yesterday.setDate(yesterday.getDate() - 1);
  const yesterdayStr = yesterday.toISOString().split("T")[0];

  const { data: yesterdayRow } = await supabase
    .from("daily_stats")
    .select("streak_maintained")
    .eq("user_id", user.id)
    .eq("date", yesterdayStr)
    .single();

  let newStreak: number;
  if (yesterdayRow?.streak_maintained) {
    newStreak = profile.current_streak + 1;
  } else {
    newStreak = 1;
  }

  const longest = Math.max(profile.longest_streak, newStreak);
  await updateProfile(supabase, { current_streak: newStreak, longest_streak: longest });
}

/** Consume a heart on wrong answer. Returns false if no hearts left. */
export async function useHeart(supabase: SupabaseClient): Promise<boolean> {
  const hearts = await rechargeHearts(supabase);
  if (hearts <= 0) return false;
  await updateProfile(supabase, { hearts: hearts - 1 });
  return true;
}

/** Get current heart count (after recharging). */
export async function getHearts(supabase: SupabaseClient): Promise<number> {
  return rechargeHearts(supabase);
}

/** Get level info for display. */
export async function getLevelInfo(
  supabase: SupabaseClient
): Promise<{ level: number; totalXp: number; xpInLevel: number; xpToNext: number }> {
  const profile = await getProfile(supabase);
  if (!profile) return { level: 1, totalXp: 0, xpInLevel: 0, xpToNext: 100 };

  const xpInLevel = profile.total_xp % 100;
  return {
    level: profile.level,
    totalXp: profile.total_xp,
    xpInLevel,
    xpToNext: 100 - xpInLevel,
  };
}
